File: 543-diameter-of-binary-tree/diameter-of-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    let d = 0
    function helper(node: TreeNode | null): number {
        if (node === null) {
            return 0;
        }
        let l = helper(node.left);
        let r = helper(node.right);
        d = Math.max(d, l + r);
        return 1 + Math.max(l, r);
    }
    helper(root);
    return d;
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    int d = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        if (root == null)
            return 0;
        height(root);
        return d;
    }

    public int height(TreeNode root) {
        if (root == null) {
            return -1;
        }
        int l = height(root.left);
        int r = height(root.right);
        d = Math.max(d, l + r + 2);
        return 1 + Math.max(l, r);
    }
}